### ISSUE TITLE
Race condition when automatically assigning IPs to VMs

### ISSUE THREAD

**Issue by vdombrovski** [apache/cloudstack#7907]
##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~
Core/Network
~~~

##### CLOUDSTACK VERSION

~~~
4.17.2
~~~

##### CONFIGURATION

N/A

##### OS / ENVIRONMENT

N/A

##### SUMMARY

When creating 2 virtual machines in the same guest network at the exact same second (in our case 23 Aug 2023 13:34:11) with automatic IP assignment, both VMs can reserve a single IP address by race condition.

##### STEPS TO REPRODUCE

With Terraform, create 2 VMs in the same Guest network

```
resource "cloudstack_instance" "terraform_vm" {
  name =  "vm-created-using-terraform"
  service_offering = "Dynamic"
  template = "Debian 11"
  network_id = resource.cloudstack_network.testnet.id
  keypair = "[name_of_ssh_keypair]"
  details = {"cpuNumber" = 4,"memory" = 2048}
  expunge = "true"
  zone = "PA4"
}

resource "cloudstack_instance" "terraform_vm_2" {
  name =  "vm-created-using-terraform-2"
  service_offering = "Dynamic"
  template = "Debian 11"
  network_id = resource.cloudstack_network.testnet.id
  keypair = "[name_of_ssh_keypair]"
  details = {"cpuNumber" = 4,"memory" = 2048}
  expunge = "true"
  zone = "PA4"
}
```

Then apply the manifest several times (until the race condition triggers), and check their IPs

##### EXPECTED RESULTS

The IPs are always different

##### ACTUAL RESULTS

In some iterations, here is what we saw (sensitive information redacted):

![acs_issue_race_condition](https://github.com/apache/cloudstack/assets/19999157/42647257-4946-4c17-9b4c-9827c8f02897)

Both VMs have reserved the exact same IP, the same Zone and guest network

Here is the relevant entries from the cloud database:

```
> select * from nics where ip4_address='10.4.1.144' and state='Allocated' \G;
*************************** 1. row ***************************
            id: 7225
          uuid: eab5163e-b88b-43a0-b6a3-ac02fd5a6037
   instance_id: 5471
   mac_address: 02:00:7c:99:00:19
   ip4_address: 10.4.1.144
       netmask: 255.255.255.0
       gateway: 10.4.1.1
       ip_type: Ip4
 broadcast_uri: NULL
    network_id: 627
          mode: Dhcp
         state: Allocated
      strategy: Start
 reserver_name: VxlanGuestNetworkGuru
reservation_id: a3a9b296-e962-49f4-8c8f-3a3eef9128da
     device_id: 0
   update_time: 2023-08-23 15:34:11
 isolation_uri: NULL
   ip6_address: NULL
   default_nic: 1
       vm_type: User
       created: 2023-08-23 13:34:11
       removed: NULL
   ip6_gateway: NULL
      ip6_cidr: NULL
  secondary_ip: 0
   display_nic: 1
*************************** 2. row ***************************
            id: 7227
          uuid: 49af630f-9075-40f3-ae95-3d5a28c19ec2
   instance_id: 5475
   mac_address: 02:00:6e:f1:00:1d
   ip4_address: 10.4.1.144
       netmask: 255.255.255.0
       gateway: 10.4.1.1
       ip_type: Ip4
 broadcast_uri: NULL
    network_id: 627
          mode: Dhcp
         state: Allocated
      strategy: Start
 reserver_name: VxlanGuestNetworkGuru
reservation_id: 5dc0a9e8-b4c5-4f8e-a883-00178984c09a
     device_id: 0
   update_time: 2023-08-23 15:34:11
 isolation_uri: NULL
   ip6_address: NULL
   default_nic: 1
       vm_type: User
       created: 2023-08-23 13:34:11
       removed: NULL
   ip6_gateway: NULL
      ip6_cidr: NULL
  secondary_ip: 0
   display_nic: 1
```

**Comment by weizhouapache** [comment]
there are some issues when deploy or start virtual machines in parellel.

for example #7202 #6644 (fixed by #6694)

**Comment by rohityadavcloud** [comment]
It seems we may have two PRs for this @sureshanaparti @DaanHoogland @weizhouapache @abh1sar and others

**Comment by DaanHoogland** [comment]
> It seems we may have two PRs for this @sureshanaparti @DaanHoogland @weizhouapache @abh1sar and others

I see only @abh1sar 's @rohityadavcloud . Can you point out the other one?

**Comment by rohityadavcloud** [comment]
^rubs my eyes^ is this same/similar https://github.com/apache/cloudstack/pull/9234 @DaanHoogland ?

**Comment by DaanHoogland** [comment]
You are right @rohityadavcloud, it is the same. I didn't see that link. @abh1sar can you contact @hsato03 on his PR and discuss a (the best) way forward

**Comment by weizhouapache** [comment]
> You are right @rohityadavcloud, it is the same. I didn't see that link. @abh1sar can you contact @hsato03 on his PR and discuss a (the best) way forward

it looks this pr aims to fix same vm IP, while #9234 aims to fix same Public Ip.

**Comment by hsato03** [comment]
I think Wei is right, they are different situations.

**Comment by abh1sar** [comment]
+1 to @weizhouapache and @hsato03 . #9234 fixes public IP race by taking lock before making changes to the IPAddressVO table.
For guest IPs the approach would be different #9240 .

**Comment by rohityadavcloud** [comment]
Thanks for checking @weizhouapache @hsato03 @abh1sar

**Comment by sureshanaparti** [comment]
Fixed in #9240
